src/main/java/at/hannibal2/skyhanni/features/inventory/caketracker/CakeTracker.kt

```diff
@@ -0,0 +1,424 @@
+import at.hannibal2.skyhanni.utils.SimpleTimeMark
+import at.hannibal2.skyhanni.utils.SkyBlockTime
+import at.hannibal2.skyhanni.utils.renderables.Renderable
+import net.minecraft.inventory.ContainerChest
+import net.minecraft.inventory.Slot
+import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
+import java.awt.Color
+import kotlin.time.Duration.Companion.milliseconds
+
+private typealias DisplayOrder = CakeTrackerDisplayOrderType
+private typealias DisplayType = CakeTrackerDisplayType
+
+@SkyHanniModule
+object CakeTracker {
+
+    private val storage get() = ProfileStorageData.profileSpecific?.cakeData
+    private val config get() = SkyHanniMod.feature.inventory.cakeTracker
+
+    private var currentYear = 0
+    private var inCakeBag = false
+    private var inCakeInventory = false
+    private var timeOpenedCakeInventory = SimpleTimeMark.farPast()
+    private var inAuctionHouse = false
+    private var auctionCakesCache = mapOf<Int, Boolean>()
+    private var searchingForCakes = false
+    private var knownCakesInCurrentInventory = listOf<Int>()
+
+    private var cakeRenderables = listOf<Renderable>()
+    private var lastKnownCakeDataHash = 0
+
+    private var unobtainedHighlightColor: Color = config.unobtainedAuctionHighlightColor.get().toChromaColor()
+    private var obtainedHighlightColor: Color = config.obtainedAuctionHighlightColor.get().toChromaColor()
+
+    /**
+     * REGEX-TEST: §cNew Year Cake (Year 360)
+     * REGEX-TEST: §cNew Year Cake (Year 1,000)
+     * REGEX-TEST: §f§f§cNew Year Cake (Year 330)
+     * REGEX-TEST: §f§f§cNew Year Cake (Year 49)
+     */
+    private val cakeNamePattern by patternGroup.pattern(
+        "cake.name",
+        "(?:§f§f)?§cNew Year Cake \\(Year (?<year>[\\d,]+)\\)",
+    )
+
+    /**
+     * REGEX-TEST: Ender Chest (2/9)
+     * REGEX-TEST: Jumbo Backpack (Slot #6)
+     * REGEX-TEST: New Year Cake Bag
+     */
+    private val cakeContainerPattern by patternGroup.pattern(
+        "cake.container",
+        "Ender Chest \\(\\d{1,2}/\\d{1,2}\\)|.*Backpack(?:§r)? \\(Slot #\\d{1,2}\\)|New Year Cake Bag",
+    )
+
+    /**
+     * REGEX-TEST: New Year Cake Bag
+     */
+    private val cakeBagPattern by patternGroup.pattern(
+        "cake.bag",
+        "New Year Cake Bag",
+    )
+
+    /**
+     * REGEX-TEST: Auctions Browser
+     * REGEX-TEST: Auctions: "Test"
+     * REGEX-TEST: Auctions: "New Year Cake (Year
+     */
+    private val auctionBrowserPattern by patternGroup.pattern(
+        "auction.search",
+        "Auctions Browser|Auctions: \".*",
+    )
+
+    /**
+     * REGEX-TEST: Auctions: "New Year C
+     */
+    private val auctionCakeSearchPattern by patternGroup.pattern(
+        "auction.cakesearch",
+        "Auctions: \"New Year C.*",
+    )
+
+    private fun addCake(cakeYear: Int) {
+        val storage = storage ?: return
+        if (cakeYear !in storage.ownedCakes) storage.ownedCakes.add(cakeYear)
+        recalculateMissingCakes()
```

Comment:
if you change ownedCakes to be a set, you can replace this with
```kt
    private fun addCake(cakeYear: Int) {
        val storage = storage ?: return
        val changed = storage.ownedCakes.add(cakeYear)
        if (changed) recalculateMissingCakes()
    }
```
same thing with removeCake